File: src/tabs/network.cpp
#include "network.hpp"

#include <algorithm>

#include "format.hpp"
#include "render.hpp"

using namespace ftxui;

static Element render_network(const AppState& s, const std::vector<SoftFork>& forks,
                              bool forks_loading) {
    auto left_col = section_box(
        "Network Status", {
                              label_value("  Active    : ", s.network_active ? "yes" : "no",
                                          s.network_active ? Color::Green : Color::Red),
                              label_value("  Peers     : ", std::to_string(s.connections)),
                              label_value("  Inbound   : ", std::to_string(s.connections_in)),
                              label_value("  Outbound  : ", std::to_string(s.connections_out)),
                          });

    auto right_col =
        section_box("Node", {
                                label_value("  Client    : ", s.subversion),
                                label_value("  Protocol  : ", std::to_string(s.protocol_version)),
                                label_value("  Relay fee : ", fmt_satsvb(s.relay_fee)),
                            });

    Elements fork_rows;
    if (forks_loading) {
        fork_rows.push_back(text("  Loading\u2026") | color(Color::GrayDark));
    } else if (forks.empty()) {
        fork_rows.push_back(text("  (unavailable \u2014 node may not support getdeploymentinfo)") |
                            color(Color::GrayDark));
    } else {
        // Header
        fork_rows.push_back(hbox({
                                text("  "),
                                text("Name") | bold | size(WIDTH, EQUAL, 18),
                                text("Type") | bold | size(WIDTH, EQUAL, 8),
                                text("Status") | bold | size(WIDTH, EQUAL, 12),
                                text("Height") | bold,
                                filler(),
                            }) |
                            color(Color::Gold1));
        fork_rows.push_back(separator());
        for (const auto& f : forks) {
            const std::string& status       = f.bip9_status.empty() ? f.type : f.bip9_status;
            Color              status_color = f.active                ? Color::Green
                                              : status == "locked_in" ? Color::Yellow
                                              : status == "started"   ? Color::Cyan
                                                                      : Color::GrayDark;
            fork_rows.push_back(hbox({
                text("  "),
                text(f.name) | size(WIDTH, EQUAL, 18),
                text(f.type) | color(Color::GrayDark) | size(WIDTH, EQUAL, 8),
                text(status) | color(status_color) | size(WIDTH, EQUAL, 12),
                f.height >= 0 ? text(fmt_height(f.height)) | color(Color::GrayDark) : text("—"),
                filler(),
            }));
        }
    }

    return vbox({
               hbox({std::move(left_col) | flex, std::move(right_col) | flex}),
               section_box("Softfork Tracking", std::move(fork_rows)),
               filler(),
           }) |
           flex;
}

NetworkTab::NetworkTab(RpcConfig cfg, Guarded<RpcAuth>& auth, ScreenInteractive& screen,
                       std::atomic<bool>& running, Guarded<AppState>& state, int refresh_secs)
    : Tab(std::move(cfg), auth, screen, running, state, refresh_secs) {}

FooterSpec NetworkTab::footer_buttons(const AppState& snap) {
    return FooterSpec{{refresh_btn(snap)}};
}

void NetworkTab::fetch() {
    if (loading_.load())
        return;
    if (thread_.joinable())
        thread_.join();
    loading_ = true;
    thread_  = std::thread([this] {
        std::vector<SoftFork> result;
        try {
            RpcClient rc(cfg_, auth_);
            auto      dep = rc.call("getdeploymentinfo")["result"]["deployments"];
            if (dep.is_object()) {
                for (const auto& [name, val] : dep.items()) {
                    SoftFork f;
                    f.name   = name;
                    f.type   = val.value("type", std::string{});
                    f.active = val.value("active", false);
                    f.height = val.value("height", int64_t{-1});
                    if (val.contains("bip9") && val["bip9"].is_object()) {
                        const auto& b9        = val["bip9"];
                        f.bip9_status         = b9.value("status", std::string{});
                        f.bip9_since          = b9.value("since", int64_t{0});
                        f.bip9_start_time     = b9.value("start_time", int64_t{0});
                        f.bip9_timeout        = b9.value("timeout", int64_t{0});
                        f.bip9_min_activation = b9.value("min_activation_height", int64_t{0});
                        if (b9.contains("statistics") && b9["statistics"].is_object()) {
                            const auto& st   = b9["statistics"];
                            f.bip9_elapsed   = st.value("elapsed", int64_t{0});
                            f.bip9_count     = st.value("count", int64_t{0});
                            f.bip9_period    = st.value("period", int64_t{0});
                            f.bip9_threshold = st.value("threshold", int64_t{0});
                        }
                    }
                    result.push_back(std::move(f));
                }
                std::sort(result.begin(), result.end(), [](const SoftFork& a, const SoftFork& b) {
                    if (a.active != b.active)
                        return a.active > b.active;
                    return a.name < b.name;
                });
            }
        } catch (...) { // NOLINT(bugprone-empty-catch)
        }
        if (!running_.load())
            return;
        softforks_ = std::move(result);
        loading_   = false;
        loaded_    = true;
        screen_.PostEvent(Event::Custom);
    });
}

Element NetworkTab::render(const AppState& snap) {
    if (!loaded_.load() && !loading_.load())
        fetch();
    std::vector<SoftFork> forks_snap;
    forks_snap = softforks_.get();
    return render_network(snap, forks_snap, loading_.load());
}

void NetworkTab::join() {
    if (thread_.joinable())
        thread_.join();
}

File: lua/tabs/network.lua
-- Network — Lua port of the built-in C++ Network tab (src/tabs/network.cpp).
--
-- Shows connection counts, node client/protocol/relay fee, and a soft-fork
-- tracking table fed by getdeploymentinfo. Built entirely from the composable
-- btcui_summary / btcui_table components so it reuses the same section-box
-- rendering and footer/refresh behaviour as the rest of the UI.
--
-- Load with: --tab lua/tabs/network.lua
-- Optional refresh interval (seconds): --tab lua/tabs/network.lua,interval=5

-- Lua 5.5 strict globals: a typo in any name below is caught at load time.
global btcui_option, btcui_rpc, btcui_set_interval, btcui_set_name,
       btcui_summary, btcui_table, ipairs, math, pairs, pcall, string, table,
       tonumber, tostring

btcui_set_name("Network")

local REFRESH = tonumber(btcui_option("interval", "5")) or 5

----------------------------------------------------------------------
-- Formatting helpers (mirror src/format.hpp)
----------------------------------------------------------------------

-- Group digits from the right with a separator, e.g. 1234567 -> 1'234'567.
local function group_digits(n, sep)
    local s   = tostring(math.floor(n))
    local out = ""
    local cnt = 0
    for i = #s, 1, -1 do
        out = s:sub(i, i) .. out
        cnt = cnt + 1
        if cnt % 3 == 0 and i > 1 then out = sep .. out end
    end
    return out
end

local function fmt_height(n) return group_digits(n, "'") end

-- relayfee is BTC/kvB; show as sat/vB (matches fmt_satsvb in src/format.hpp).
local function fmt_satsvb(btc_per_kvb)
    return string.format("%.1f sat/vB", (btc_per_kvb or 0) * 1e5)
end

local function yesno(v, yes_color, no_color)
    return { value = v and "yes" or "no", color = v and yes_color or no_color }
end

----------------------------------------------------------------------
-- Panels — Network Status | Node side by side, Softfork Tracking below
-- (two consecutive summaries render side-by-side; a table flushes the run
--  and starts a new full-width row, mirroring the C++ tab's layout)
----------------------------------------------------------------------

local status_panel = btcui_summary({
    title = "Network Status",
    fields = {
        { name = "active",   label = "Active" },
        { name = "peers",    label = "Peers" },
        { name = "inbound",  label = "Inbound" },
        { name = "outbound", label = "Outbound" },
    },
})

local node_panel = btcui_summary({
    title = "Node",
    fields = {
        { name = "client",   label = "Client" },
        { name = "protocol", label = "Protocol" },
        { name = "relayfee", label = "Relay fee" },
    },
})

local fork_table = btcui_table({
    key = "name",
    title = "Softfork Tracking",
    columns = {
        { name = "name",   header = "Name" },
        { name = "type",   header = "Type" },
        { name = "status", header = "Status" },
        { name = "height", header = "Height" },
    },
})

----------------------------------------------------------------------
-- Refresh
----------------------------------------------------------------------

-- Status colour mirrors render_network() in network.cpp:
--   active -> green, locked_in -> yellow, started -> cyan, else -> gray
local function status_color(active, status)
    if active               then return "green" end
    if status == "locked_in" then return "yellow" end
    if status == "started"   then return "cyan" end
    return "gray"
end

local function refresh()
    -- Network Status + Node panels (from getnetworkinfo).
    -- The C++ tab read these from the shared AppState populated by the global
    -- poll thread; a Lua tab has no access to AppState, so we fetch directly —
    -- the same approach dashboard.lua takes.
    local oknet, net = pcall(btcui_rpc, "getnetworkinfo")
    if oknet and net then
        status_panel:set({
            active   = yesno(net.networkactive, "green", "red"),
            peers    = tostring(net.connections or 0),
            inbound  = tostring(net.connections_in or 0),
            outbound = tostring(net.connections_out or 0),
        })
        node_panel:set({
            client   = net.subversion or "",
            protocol = tostring(net.protocolversion or 0),
            relayfee = fmt_satsvb(net.relayfee),
        })
    end

    -- Softfork tracking (from getdeploymentinfo).
    local ok, info = pcall(btcui_rpc, "getdeploymentinfo")
    local dep = (ok and info and info.deployments) or nil

    fork_table:start_refresh()
    if dep then
        -- Collect deployments into an array so we can sort them the same way
        -- the C++ tab does: active first, then by name ascending.
        local forks = {}
        for name, val in pairs(dep) do
            local b9     = val.bip9 or {}
            local status = (b9.status and b9.status ~= "") and b9.status or (val.type or "")
            forks[#forks + 1] = {
                name   = name,
                type   = val.type or "",
                status = status,
                active = val.active == true,
                height = val.height,
            }
        end
        table.sort(forks, function(a, b)
            if a.active ~= b.active then return a.active end
            return a.name < b.name
        end)
        for _, f in ipairs(forks) do
            local height_str = (f.height and f.height >= 0) and fmt_height(f.height) or "—"
            fork_table:update(f.name, {
                name   = f.name,
                type   = { value = f.type,   color = "gray" },
                status = { value = f.status, color = status_color(f.active, f.status) },
                height = { value = height_str, color = "gray" },
            })
        end
        -- Clear any previous "unavailable" banner now that we have data.
        fork_table:set_header_info("")
    else
        fork_table:set_header_info({
            value = "(unavailable — node may not support getdeploymentinfo)",
            color = "gray",
        })
    end
    fork_table:finish_refresh()
end

btcui_set_interval(REFRESH, refresh)
